tools/codegen/gen.py

```diff
@@ -176,6 +176,94 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
         return f'm.def({cpp_string(str(f.func))});\n'
 
 
+def _num_leading_spaces(line: str) -> int:
+    return len(line) - len(line.lstrip())
+
+
+# Unindents all lines in code. Each line gets unindented the same amount;
+# that amount is equal to the smallest number of leading spaces across all lines
+def deindent(code: str) -> str:
+    lines = code.split('\n')
+    min_leading_spaces = min(map(_num_leading_spaces, lines))
+    lines = [line[min_leading_spaces:] for line in lines]
+    return '\n'.join(lines)
+
+
+# Generates UnambiguousTorchOperators.h and UnambiguousTorchOperators.cpp.
+# These provide macros that, given an operator and overload name, allow users
+# to access an "un-overloaded" function version of the operator. This
+# is useful for extension writers who want to (1) want to decltype the operator
+# and (2) don't want to worry about method-only operators.
+@dataclass(frozen=True)
+class UnambiguousTorchOperators:
+    target: Union[
+        Literal[Target.DECLARATION],
+        Literal[Target.DEFINITION]
+    ]
+
+    @method_with_native_function
+    def __call__(self, f: NativeFunction) -> Optional[str]:
+        unambiguous_name = self.unambiguous_function_name(f)
+        # NB: requires_grad is the only exception to the rule because
+        # its const correctness is questionable.
+        if unambiguous_name in set(['requires_grad_']):
+            return None
+
+        if self.target is Target.DECLARATION:
+            return self.gen_declaration(f)
+
+        if self.target is not Target.DEFINITION:
+            assert_never(self.target)
+
+        return self.gen_definition(f)
+
+    # Given a function schema "aten::op.overload(...)",
+    # If there is no overload name, this returns f"{op}"
+    # If there is an overload name, this returns f"{op}_{overload}"
+    def unambiguous_function_name(self, f: NativeFunction) -> str:
+        base_name = str(f.func.name.name)
+        overload_name = f.func.name.overload_name
+        if overload_name:
+            return f'{base_name}_{overload_name}'
+        return base_name
+
+    def gen_declaration(self, f: NativeFunction) -> str:
+        unambiguous_name = self.unambiguous_function_name(f)
+        returns_type = dispatcher.returns_type(f.func.returns).cpp_type_registration_declarations()
```

Comment:
You can use the signature classes in `types.py`, plus `translate()` to do most of the work for you when generating the declaration/definitions and translating from the dispatcher API to C++ API. Taking some inspiration from [`register_dispatch_key.py`](https://github.com/pytorch/pytorch/blob/ff982ef73d91bf384d8f6df65a0a63dfbc998b7d/tools/codegen/dest/register_dispatch_key.py#L149), something like:

```
unambiguous_name = self.unambiguous_function_name(f)
sig = DispatcherSignature.from_schema(f)

if self.target is Target.Declaration:
    return f'TORCH_API {sig.decl(unambiguous_name)};'
elif self.target is Target.Definition:
    target_sig = CppSignatureGroup.from_native_function(f, method=False).most_faithful_signature()
    return f"""
{sig.defn(unambiguous_name)} {{
    return {target_sig.name()}({', '.join(e.expr for e in translate(sig.arguments(), target_sig.arguments()))});
}}"""
    else:
        assert_never(self.target)
```

I'm pretty sure that `CppSignatureGroup.from_native_function(f, method=False)` should do all the work of dealing with method-only ops for you.

Reply:
Thanks for the tip! DispatcherSignature.decl()/defn() work and make this a lot simpler.

Is there a utility function somewhere that handles writing invocations for method-only operators (e.g. "at::sin(self)" vs "self.sin_()")? I couldn't find one so I rolled my own in the updated code